cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java

```diff
@@ -874,47 +932,104 @@ private void copyVolumes(String namespace, String cdapNamespace) throws IOExcept
   }
 
   /**
-   * Copy service account into the new namespace for deployments created via the KubeTwillRunnerService
+   * Create service account and role bindings required for workload pod.
    * TODO: (CDAP-18956) improve this logic to be for each pipeline run
    */
-  private void copyServiceAccount(String namespace, String cdapNamespace, String accountName) throws IOException {
+  private void createWorkloadServiceAccount(String namespace, String cdapNamespace) throws IOException {
+    try {
+      // Create service account for workload pod
+      // TODO(CDAP-19149): Cleanup strong coupling currently present in CDAP service accounts to avoid copying.
+      String serviceAccountName = podInfo.getServiceAccountName();
+      createServiceAccount(namespace, cdapNamespace, serviceAccountName);
+
+      // Create namespace-specific role-binding for the workload service account
+      createNamespacedRoleBinding(WORKLOAD_LAUNCHER_NAMESPACE_ROLE_BINDING_NAME, CLUSTER_ROLE_KIND,
+                                  workloadLauncherRoleNameForNamespace, namespace, serviceAccountName, cdapNamespace);
+
+      // Create cluster-wide role-binding for the workload service account
+      String workloadLauncherClusterRoleBindingName = String.format(WORKLOAD_LAUNCHER_CLUSTER_ROLE_BINDING_FORMAT,
```

Comment:
does k8s limit how long these can be or the character set that can be used? Wondering if we need to sanitize the namespace and if so, how to handle conflicts.

Reply:
Kubernetes expects role-bindings and cluster role-bindings names to be in [Path Segment Name](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#path-segment-names) format which means it cannot be `.` and `..` and cannot contain the `/` and `%` characters. I've gone ahead and added validation to the name to account for this.

Reply:
does this mean it will just always fail for some namespaces? Or do namespace also have those restrictions? (in which case we don't need to do the validation checks)

Reply:
Namespace names must always be [RFC-1123 compliant](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/#namespaces-and-dns):
 - contain at most 63 characters
 - contain only lowercase alphanumeric characters or '-'
 - start with an alphanumeric character
 - end with an alphanumeric character

As such, it's guaranteed to be valid for a namespace role binding given the fact that it's technically a subset of the path segment name validation. It might be worth adding validation for namespace names as well since CDAP namespaces can be upper-case, but k8s namespaces must all be lower-case.

Reply:
there's validation to check that the kube namespace meets the requirements https://github.com/cdapio/cdap/blob/develop/cdap-kubernetes/src/main/java/io/cdap/cdap/master/environment/k8s/KubeMasterEnvironment.java#L206

Reply:
Thanks for pointing this out! Moved that check to the utility file and added some unit tests for consistency.